compiler/rustc_data_structures/src/sync.rs

```diff
@@ -92,6 +93,37 @@ mod mode {
 
 pub use mode::{is_dyn_thread_safe, set_dyn_thread_safe_mode};
 
+/// A guard used to hold panics that occur during a parallel section to later by unwound.
+/// This is used for the parallel compiler to prevent fatal errors from non-deterministically
+/// hiding errors by ensuring that everything in the section has completed executing before
+/// continuing with unwinding. It's also used for the non-parallel code to ensure error message
+/// output match the parallel compiler for testing purposes.
+pub struct ParallelGuard {
+    panic: Lock<Option<Box<dyn Any + std::marker::Send + 'static>>>,
```

Comment:
Should we implement `Drop` for `ParallelGuard` to make sure we never drop a panic on the floor without handling it? `unwind` will need to get reworked below if so.

Reply:
I initially went for `Drop`, but I ended up preferring the explicit unwinding of the `unwind` calls.

Reply:
No, I think having the explicit unwind calls is better. I just wonder if we should protect against missing unwind calls somehow... anyways, I don't feel particularly strongly about it.